// supabase-wechat-stable/src/storage-js/src/lib/fetch.ts
export type Fetch = typeof fetch

export interface FetchOptions {
  header?: {
    [key: string]: string
  }
  noResolveJson?: boolean
}

export interface FetchParameters {
  signal?: AbortSignal
}

export type RequestMethodType = 'GET' | 'POST' | 'PUT' | 'DELETE'

const _getErrorMessage = (err: any): string =>
  err.msg || err.message || err.error_description || err.error || JSON.stringify(err)

const handleError = (error: any, reject: any) => {
  if (typeof error.json !== 'function') {
    return reject(error)
  }
  error.json().then((err: any) => {
    return reject({
      message: _getErrorMessage(err),
      status: error?.status || 500,
    })
  })
}

const _getRequestParams = (
  method: RequestMethodType,
  options?: FetchOptions,
  parameters?: FetchParameters,
  data?: object
) => {
  const params: { [k: string]: any } = { method, headers: options?.header || {} }

  if (method === 'GET') {
    return params
  }

  params.headers = { 'Content-Type': 'application/json', ...options?.header }
  params.data = JSON.stringify(data)
  return { ...params, ...parameters }
}
let _fetch = require('wefetch')
async function _handleRequest(
  method: RequestMethodType,
  url: string,
  options?: FetchOptions,
  parameters?: FetchParameters,
  data?: object
): Promise<any> {
  return new Promise((resolve, reject) => {
    let wfRequestParams = _getRequestParams(method, options, parameters, data)
    wfRequestParams = {
      ...wfRequestParams,
      header: wfRequestParams.headers,
      data: JSON.parse(wfRequestParams.body),
    }
    _fetch(url, wfRequestParams)
      .then((result: any) => {
        if (result.statusCode == 200) throw result
        if (options?.noResolveJson) return resolve(result)
        return result
      })
      .then((data: any) => resolve(data))
      .catch((error: any) => handleError(error, reject))
  })
}

export async function get(
  url: string,
  options?: FetchOptions,
  parameters?: FetchParameters
): Promise<any> {
  return _handleRequest('GET', url, options, parameters)
}

export async function post(
  url: string,
  data: object,
  options?: FetchOptions,
  parameters?: FetchParameters
): Promise<any> {
  return _handleRequest('POST', url, options, parameters, data)
}

export async function put(
  url: string,
  data: object,
  options?: FetchOptions,
  parameters?: FetchParameters
): Promise<any> {
  return _handleRequest('PUT', url, options, parameters, data)
}

export async function remove(
  url: string,
  data: object,
  options?: FetchOptions,
  parameters?: FetchParameters
): Promise<any> {
  return _handleRequest('DELETE', url, options, parameters, data)
}
